corehq/apps/domain/models.py

```diff
@@ -957,7 +957,7 @@ def email_to_request(self):
 
         send_html_email_async.delay(
             _('Transfer of ownership for CommCare project space.'),
-            self.to_user.email,
+            self.to_user.get_email(),
```

Comment:
What's the difference between the `email` attribute and `get_email()`? I would prefer to see us use `user.email` rather than `user.get_email()` if they are the same since it's shorter and more idiomatic Python. If they are different can we override the `email` attribute with a property that calls `get_email()`?

Getters are so Java, yo, but maybe you like enterprise beans? /s

Reply:
i believe `get_email()` defaults to `username` if `user.email` is not specified, otherwise it returns `user.email`

Reply:
Yeah what Biyeun said is correct. The email field isn't required so there is no guarantee it exists. This does beg the question, why isn't an email field required? I don't see a good reason to allow a user to set it to None because we will still try to reach out via their username, and there are certain communications that a user shouldn't be able to "unsubscribe" from.

Reply:
@millerdev following up on:
> If they are different can we override the email attribute with a property that calls get_email()?

I agree it would be cleaner to not have a getter, but my understanding is that we can't override the actual `email` property with a `@property` decorated method named `email`, since it would be ambiguous which property to refer to? So as long as the model properties are named `email`, we need a getter named something else, and `get_email` seems like the next best thing unfortunately. Let me know if I'm missing something though.

Reply:
I was worried it would be that way. It would be nice if the email model field were named `_email`, leaving room for an `email` property that uses `get_email` and a setter that sets `_email`. We could implement that if we felt it was worth it (even without renaming any database columns), but fine to leave that for a future enhancement.

My main concern is that people who are not familiar with the difference will accidentally use the `email` attribute when they should be using `get_email()`. "There should be one-- and preferably only one --obvious way to do it."

Reply:
>My main concern is that people who are not familiar with the difference will accidentally use the email attribute when they should be using get_email(). "There should be one-- and preferably only one --obvious way to do it."


Absolutely agree. I was intending to ask a wider audience how we disseminate this kind of information to all devs, but after this conversation realize the best possible solution is make it easy to use as a caller of this, so updating this to rely on `email` instead of `get_email()` does sound nice. 

>even without renaming any database columns

How would you go about this?

Reply:
I was thinking these were SQL model columns but on further investigation I see they are `DocumentSchema` columns. That requires a different approach than what I was originally thinking. The first thing that comes to mind is change this line

https://github.com/dimagi/commcare-hq/blob/6f865e021762ac20779894f521a5a3ecc0934bb9/corehq/apps/users/models.py#L710

to use a custom subclass of `LowercaseStringProperty` that overrides [`__get__`](https://github.com/dimagi/jsonobject/blob/61fed25f1dbe9bf231ca2897d1b80b4dfee615b1/jsonobject/base_properties.pyx#L71). Something like
```py
class UserEmailProperty(LowercaseStringProperty):

    def __get__(self, instance, owner):
        if instance is None:
            return self
        email = super().__get__(instance, owner)
        return email or instance.get("username")
```

Reply:
Offlined with Daniel, and agree that it is worth implementing something to replace the `get_email()` design for users. We acknowledged that the `username` field is only a valid email for a WebUser, so this override would need to either be aware of the doc type it is dealing with, or check that the `username` is a valid email before returning it. Another potential solution is to override a property type in a document subclass, but that requires more investigation to see if it is possible. I'm going to do this on a separate PR since this is a related, but separate issue.